crates/storage/store_db/libmdbx.rs

```diff
@@ -82,6 +125,62 @@ impl Store {
     ) -> Result<Option<BlockHash>, StoreError> {
         Ok(self.read::<CanonicalBlockHashes>(number)?.map(|a| a.to()))
     }
+
+    fn add_block_header_with_txn(
+        block_hash: BlockHash,
+        block_header: BlockHeader,
+        write_txn: &mut libmdbx::orm::Transaction<'_, libmdbx::RW>,
+    ) -> Result<(), StoreError> {
+        Self::write_with_txn::<Headers>(write_txn, block_hash.into(), block_header.into())
+    }
+
+    fn add_block_number_with_txn(
+        block_hash: BlockHash,
+        block_number: BlockNumber,
+        write_txn: &mut libmdbx::orm::Transaction<'_, libmdbx::RW>,
+    ) -> Result<(), StoreError> {
+        Self::write_with_txn::<BlockNumbers>(write_txn, block_hash.into(), block_number)
+    }
+
+    fn add_block_body_with_txn(
+        block_hash: BlockHash,
+        block_body: BlockBody,
+        write_txn: &mut libmdbx::orm::Transaction<'_, libmdbx::RW>,
+    ) -> Result<(), StoreError> {
+        Self::write_with_txn::<Bodies>(write_txn, block_hash.into(), block_body.into())
+    }
+
+    fn add_transaction_locations_with_txn(
+        locations: Vec<(H256, BlockNumber, BlockHash, Index)>,
+        write_txn: &mut libmdbx::orm::Transaction<'_, libmdbx::RW>,
+    ) -> Result<(), StoreError> {
+        #[allow(clippy::type_complexity)]
+        let key_values = locations
+            .into_iter()
+            .map(|(tx_hash, block_number, block_hash, index)| {
+                (tx_hash.into(), (block_number, block_hash, index).into())
+            });
+
+        Self::write_batch_with_txn::<TransactionLocations>(write_txn, key_values)
+    }
+
+    fn add_transaction_locations(
+        &self,
+        locations: Vec<(H256, BlockNumber, BlockHash, Index)>,
+    ) -> Result<(), StoreError> {
+        self.write_with_closure(|txn| Self::add_transaction_locations_with_txn(locations, txn))
+    }
+
+    pub fn add_block_body_and_header(&self, block: Block) -> Result<(), StoreError> {
+        let header = block.header;
+        let number = header.number;
+        let hash = header.compute_block_hash();
+        self.write_with_closure(|txn| {
+            Self::add_block_body_with_txn(hash, block.body, txn)?;
+            Self::add_block_header_with_txn(hash, header, txn)?;
+            Self::add_block_number_with_txn(hash, number, txn)
+        })
+    }
```

Comment:
Ahhh this looks better. I missed this one. I think it's still missing the caller.